ifrs17/Report/ReportMutableScopesInteractive.ipynb

```diff
@@ -0,0 +1,1094 @@
+                "\n        if(InputDataCube is null) return Enumerable.Empty<ReportVariable>().ToDataCube();",
+                "\n        var filteredDataCube = (Identity.dataFilter is null || Identity.dataFilter.Length == 0) ? InputDataCube : InputDataCube.Filter(Identity.dataFilter); ",
+                "\n        if(Identity.scenario == null) return filteredDataCube.Where(x => x.Scenario == null).ToDataCube();",
+                "\n        if(Identity.scenario != \"All\" && Identity.scenario != \"Delta\") return filteredDataCube.Filter((\"Scenario\", Identity.scenario));",
+                "\n        if(Identity.scenario == \"All\") return filteredDataCube.Select(x => x.Scenario == null ? x with {Scenario = \"Best Estimate\" } : x).ToDataCube();",
+                "\n        var bestEstimateById = filteredDataCube.Where(x => x.Scenario == null).ToDictionary(x => x.ToIdentityString());",
+                "\n        return filteredDataCube.Select(x => x.Scenario == null ? x with { Scenario = \"Best Estimate\" } ",
+                "\n                                                : x with { Value = x.Value - (bestEstimateById.TryGetValue((x with {Scenario = null}).ToIdentityString(), out var be)? be.Value : 0.0) }).ToDataCube();",
+                "\n    }}",
+                "\n} "
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "[InitializeScope(nameof(Init))]",
+                "\npublic interface ReportScope : IMutableScope<string>, ReportingNodeFormsEntity, MonthlyPeriodFormsEntity, ScenarioFormsEntity, CurrencyFormsEntity, ",
+                "\n                               BasicSliceAndDiceRowsFormsEntity, BasicSliceAndDiceColumnsFormsEntity, BasicFilterFormsEntity {",
+                "\n    protected IPivotFactory report => GetStorage().Report;",
+                "\n    protected IExportVariable export => GetStorage().Export;",
+                "\n    protected int headerColumnWidthValue => 250;",
+                "\n",
+                "\n    HashSet<(ReportIdentity, CurrencyType)> GetDataIdentities() => GetStorage().GetIdentities((Year, Month), ReportingNode, Scenario, CurrencyType); // TODO, add filter for identities, if the property is exposed at this level",
+                "\n",
+                "\n    IDataCube<ReportVariable> GetData() => default;",
+                "\n",
+                "\n    async Task<GridOptions> ToReportAsync() {",
+                "\n        await GetStorage().InitializeAsync((Year, Month), ReportingNode, Scenario, CurrencyType);",
+                "\n        var dataScope = GetScope<Data>(((Year, Month), ReportingNode, Scenario, CurrencyType, dataFilter));",
+                "\n        var dataCube = GetData();",
+                "\n        // Find another way to avoid nasty errors if the DC is empty - AK",
```

Comment:
Use Any()? And do we need this?

Reply:
Changed this to Any(). I would prefer to have it, it avoids spitting out an error when there is no data to show.

Reply:
OK, I will a comment here and @dcolleoni will put it later on the developers board.